GVS PGEN Extract

KevinCLydon opened the pull request:
# Extracting PGEN from GVS

## The PGEN format
PGEN is a format written for and used by version 2 of [PLINK](https://www.cog-genomics.org/plink/2.0/).  ***IT IS VERY IMPORTANT TO NOTE THAT VERSION 2 OF PLINK IS STILL IN ALPHA AND THE PGEN FORMAT IS STILL SUBJECT TO CHANGE.***  The format comprises 3 file types (or actually sometimes 4):
1. A `.pgen` file.  This is a binary file that stores a mapping of samples and sites to genotypes in a very cleverly compressed way that I can't explain super well because it's complicated.
2. A `.pvar` file. This is essentially a sites-only VCF.  It has information for each site referenced in the `.pgen` file.  PLINK also provides an option to produce/use a zstd compressed version of the file (`.pvar.zst`), and we have opted to write that for performance purposes.
3. A `.psam` file.  This is a plaintext file that contains a list of samples referenced in the `.pgen` file.  It also optionally includes some phenotype data.
4. Optionally, a `.pgi` file.  Typically, a `.pgen` file has an index at the top.  Optionally, PLINK supports using a `.pgen` file with an index in a separate `.pgi` file.

The PGEN format does not store all of the information that a VCF has.  It leaves out a lot of the fields and annotations you can store in a VCF.  As a result of this and the clever compression in the `.pgen` file, these files are typically much smaller than equivalent VCFs.

For more information on the PGEN file format, see the official spec [here](https://github.com/chrchang/plink-ng/blob/master/pgen_spec/pgen_spec.pdf).

## The code
The code for the PGEN extract can be divided into 3 parts:
1. The PGEN-JNI, a C++/JNI library that handles writing HTSJDK VariantContext objects to PGEN files,
2. ExtractCohortToPgen, a GATK tool based on ExtractCohortToVcf that processes VariantContexts and passes them to PGEN-JNI for writing, and
3. GvsExtractCallsetPgenMerged, a WDL workflow based on GvsExtractCallset that uses ExtractCohortToPgen to write a series of PGEN files and then merges them by chromosome.

### Part 1: PGEN-JNI
The PGEN-JNI library was written by Chris Norman of the GATK Engine Team and lives [here](https://github.com/broadinstitute/pgen-jni).  It is written primarily in C++ for performance purposes and also compatibility with the pgenlib library (part of the [plink repo](https://github.com/chrchang/plink-ng/tree/master)).  It builds on top of pgenlib to provide a writer for creating PGEN files and writing to them from HTSJDK VariantContext objects.

PGEN-JNI is compatible with Linux and macOS.

A build of this library is currently hosted on the Broad's artifactory repo, and that is being used as a dependency for GATK.

Ownership of the PGEN-JNI library will stay with the GATK Engine Team and we will provide support for it if y'all encounter any issues with it.

### Part 2: ExtractCohortToPgen
ExtractCohortToPgen is a GATK tool that inherits from ExtractCohort and is based very closely on ExtractCohortToVcf.  It produces 3-4 files:

1. A `.pgen` file, which contains a mapping of samples and sites to variants,
2. A `.psam` file, which contains a list of sample names,
3. A `.pvar.zst` file, which is a zstd compressed list of sites with alleles, similar to a sites-only VCF, and
4. Optionally (if specified by setting `write-mode` to `WRITE_SEPARATE_INDEX`), a `.pgi` file, which contains an index for the `.pgen` file.

It has a few arguments that are specific to it that warrant explanation.

#### pgen-chromosome-code
Plink defines a set of [chromosome codes](https://www.cog-genomics.org/plink/2.0/data#irreg_output) that correspond to different sets of contig names of chromosomes.  This tool supports two of those chromosome code options: `chrM` and `MT`, which correspond to hg38 and hg19 contig naming, respectively.  This argument is required.

#### write-mode
The PGEN writer defined in PGEN-JNI defines two different write modes.  When `WRITE_AND_COPY` is selected, a temporary .pgen file is created and written to during the running of the tool, and then once all records have been written, a new file is created with the index at the top and the contents of the temporary .pgen file appended to it.  When `WRITE_SEPARATE_INDEX` is selected, the index is instead written to a separate .pgi file.  The default is `WRITE_AND_COPY`.

#### max-alt-alleles
The PGEN format can only support up to 254 alt alleles per site.  This argument allows you to specify a limit.  The default is the max of 254.  Any sites with more alt alleles than the specified max will not be written.

#### lenient-ploidy-validation
PGEN is a bit quirky in that it requires samples to be diploid but has a special case for sex chromosomes, which are allowed to be haploid.  By default, any attempt to write a record with an unsupported ploidy will result in an exception being thrown.  If this flag is used, then ploidy failures will instead be logged and the records will be written as missing.

#### writer-log-file
The C++ code in the PGEN writer in PGEN-JNI will log sites that exceed max-alt-alleles and with unsupported ploidy (if lenient-ploidy-validation is set) to the specified log file, if this argument is set.

#### allow-empty-pgen
Empty PGEN files are not technically valid PGEN files.  However, for parallel processing purposes, it is sometimes helpful to allow the creation of empty files when there are no variants to be written.  The GvsExtractCallsetPgenMerged workflow relies on this.

If this flag is set and no variants are written, an empty .pgen, .psam, and .pvar.zst file will be written in `onShutdown()`.

By default (i.e. if this flag is not set), if there are no variants written, an exception will be thrown.  

### Part 3: GvsExtractCallsetPgenMerged
GvsExtractCallsetPgenMerged is a WDL workflow that calls ExtractCohortToPgen to extract data from GVS and write it to PGEN files, and then merges those PGEN files by chromosome.  This workflow has 3 steps:

#### Step 1: GvsExtractCallsetPgen
This is a workflow based very closely on the GvsExtractCallset workflow (which is used for extracting data from GVS and writing it to VCF files).  It has a few notable differences when compared to that workflow.

First, it features extra inputs for some of the PGEN-specific arguments described in [Part 2](#part-2-extractcohorttopgen).  These mostly have defaults which match the defaults for the ExtractCohortToPgen tool.  A default of "chrM" is also specified for chromosome code, to match the reference defaulting to hg38.

Second, the call to SplitIntervals is replaced with a call to a new task called SplitIntervalsTarred.  For very large scatter counts (I think >20k-ish?), the SplitIntervals task will fail because the shell command being run behind the scenes by the WDL `glob()` function will fail when trying to do wildcard expansion with that many files.  The best solution I was able to find for this was to just tar up all the interval files and use that as an output.

Third, as a result of this, the ExtractTask has some changes when compared to the VCF ExtractTask (beyond the fact that it now calls the PGEN extract tool instead of the VCF one).  It now accepts the interval lists tar file as an input, along with the filename for the interval list file for that particular shard.  Before it runs the extract tool, it extracts the specified interval list file from the interval lists tar.

The output of this workflow is a list of .pgen, .psam, and .pvar.zst files, along with the interval list tar and a list of the filenames within the tar (ordered to match the pgen/psam/pvar files so they can be matched up to each other more easily).

#### Step 2: SplitFilesByChromosome
This is a short task defined within GvsExtractCallsetPgenMerged.  It loops through the interval list files in the interval list tar and extracts the contig name from each.  Then, going by that contig name, it writes the names of the corresponding .pgen, .psam, and .pvar.zst files to list files named for the corresponding contigs.  

The output of this task is 3 arrays of files, containing names for .pgen, .psam, and .pvar.zst files which contain sites for each contig.  For example, the file `chr1.pgen_list` would contain a list of all the .pgen files that contain variants for sites on chromosome 1.

#### Step 3: MergePgenHierarchical
This workflow accepts a list of .pgen, .psam, and .pvar.zst files and merges them all into one file using Plink's `--pmerge-list` functionality.

For performance purposes, the merging is done in two stages.  One big, monolithic merge takes too long.

First, the file lists are sorted by index (the index in the filename).  Plink does not like (i.e. does not support) merging files with overlapping intervals, so the files need to be sorted so that when they are merged in stages, we do not create merged files with intervals that overlap.

Next, the lists are split, so we can parallelize the merge.  Each of the split lists is then sent to a task that merges the files using Plink.

Then, this is repeated to merge all the resulting files into one.  The output is a single .pgen file, .psam file, and .pvar.zst file containing all the extracted data for the corresponding chromosome.

## Testing

### Unit tests
Unit tests for the ExtractCohortToPgen tool exist in the ExtractCohortToPgenTest file.  They are based closely on the tests in ExtractCohortToVcfTest, with a few extras to account for behaviors specific to PGEN extract.

### Large scale testing
All of my testing (with the exception of some very small scale stuff early on) has been done in the [GVS_AoU_PGEN_Extract_Development Terra workspace](https://app.terra.bio/#workspaces/allofus-drc-prod-auxiliary/GVS_AoU_PGEN_Extract_Development) and using data from the GVS Delta callset (aou-genomics-curation-prod.aou_wgs_fullref_v2).  My test process (for the majority of tests) has been as follows:

1. Select a list of sample names from aou_wgs_fullref_v2.sample_info (excluding control samples and sample 3224672 because of the data issue mentioned [here](https://broadinstitute.slack.com/archives/CJRLP6ZSA/p1699026273329339)).
2. Use that list as an input to GvsPrepareRangesCallset to create a cohort of test data in a separate BigQuery dataset (aou-genomics-curation-prod.klydon_pgen_extract_test).
3. Run GvsExtractCallsetPgen on the newly created cohort.  (I would just run GvsExtractCallsetPgenMerged, but I like using Workflow Dashboard to monitor how the job is going and dig into it if there are any failures.  Workflow Dashboard doesn't seem to let you dig into individual tasks for workflows with sub-workflows, so it wouldn't allow me to look at individual shards running ExtractTask if I ran GvsExtractCallsetPgenMerged.  Job Manager would be an alternative for this, but it seems to be pretty much unusable for even moderately-sized jobs.)
4. Run GvsExtractCallset on the newly created cohort, making sure to use the same parameters, including scatter count.  This will generate VCF files that we can use to compare to the PGEN files created during the previous step for validation.
5. Run GvsExtractCallsetPgenMerged with the same parameters used to run GvsExtractCallsetPgen in Step 3.  This will use call-caching for the extract steps and then merge the PGEN files by chromosome.  (Running it this way is maybe not the ideal way to do this, but it's what I've been doing for reasons described in the parenthetical in Step 3).
6. Create list files, by file type, containing the gs:// URIs for the .pgen, .psam, and .pvar.zst files created in Step 3, along with the .vcf.gz files created in Step 4.  Upload them to the workspace to use for validation.
7. Run ComparePgenAndVcfScatter using the file lists as inputs.  If there are any differences, it will output files that contain those differences.  If there are no diff files generated, the files match.

ComparePgenAndVcfScatter is a workflow I wrote that converts a list of .pgen, .psam, and .pvar.zst files generated by GvsExtractCallsetPgen into .vcf.gz files and then compares those files to a list of .vcf.gz files generated by GvsExtractCallset.  

It ignores basically everything except genotypes, because PGENs do not store all the other fields and annotations that the VCFs might have.  It will also skip over any sites in the VCFs with >254 alleles because those will not be present in the PGEN files.  Any differences are written to diff files, in the form of the differing lines in the VCFs being compared.

The code for this comparison tool lives [here](https://github.com/KevinCLydon/pgen_vcf_comparator) in a repo I created under my GitHub account.  (I didn't create it under the Broad org because it's sort of half-baked and bad and not actually meant to be used by anyone other than me.)  I don't know if y'all want to continue using this tool, but I'm happy to discuss it more if it would actually be useful to you.

## To-dos / caveats

### PGEN-JNI
The version of PGEN-JNI I'm referencing in the current build.gradle file is a beta version that is hosted on artifactory.  Functionally, this is totally fine, but we want to get a 1.0 version of it hosted publicly.  Chris Norman, who developed the tool is currently very working on this and very close to done.  Once he's completed this, I want to run a sanity test or two against a small subset of the Delta callset just to make sure everything is functioning as intended.

### Merging by chromosome arm
Right now, the last step of the PGEN extract workflow merges the PGEN files by contig name, so the final result is one trio of files (.pgen, .psam, and .pvar.zst) per chromosome.  There was discussion about changing this to merge instead by chromosome arm.  I want to make this change, but it's not super simple, so I've prioritized getting this version of the code ready for merging before tackling that.

### The PGEN format
As I mentioned above, Plink 2.0 and the PGEN file format are still not in full release, so the format could be subject to change in the future, which will require updates to our PGEN writing code and could possibly introduce problems.

mcovarr commented on Dockerfile:
```diff
@@ -80,6 +80,13 @@ RUN echo "source activate gatk" > /root/run_unit_tests.sh && \
     echo "ln -s /gatkCloneMountPoint/build/ /gatkCloneMountPoint/scripts/docker/build" >> /root/run_unit_tests.sh && \
     echo "cd /gatk/ && /gatkCloneMountPoint/gradlew -Dfile.encoding=UTF-8 -b /gatkCloneMountPoint/dockertest.gradle testOnPackagedReleaseJar jacocoTestReportOnPackagedReleaseJar -a -p /gatkCloneMountPoint" >> /root/run_unit_tests.sh
 
+# TODO: Determine whether we actually need this.  For now it seems to be required because the version of libstdc++ on
+# TODO: the gatk base docker is out of date (maybe?)
```
Our `ah_var_store` branch is out of date with `master` again; is it possible that this issue would go away once we refresh?

KevinCLydon replied:
It might, yeah.  I know the base image in master was recently updated to Ubuntu 22.04, which might have the more recent version of the necessary library.

KevinCLydon replied:
I know having this step in the docker resolves the issue, so I'd say it's very low priority.  We could probably just remove it and run a quick test when the refresh happens.

mcovarr commented on scripts/variantstore/wdl/GvsExtractCallsetPgen.wdl:
```diff
@@ -0,0 +1,603 @@
+        # If false, the output files will be named with a non-zero-padded interval number suffix
+        Boolean zero_pad_output_pgen_filenames = true
+
+        # set to "NONE" if all the reference data was loaded into GVS in GvsImportGenomes
```
Which btw it is not for Echo, we dropped GQ 0

KevinCLydon replied:
Oh that's interesting.  Should the default value be changed/removed in that case?

mcovarr commented on scripts/variantstore/wdl/GvsExtractCallsetPgen.wdl:
```diff
@@ -0,0 +1,603 @@
+                                                 else if GetNumSamplesLoaded.num_samples < 20000 then 2000 # Stroke Anderson
+                                                      else if GetNumSamplesLoaded.num_samples < 50000 then 10000
+                                                           else if GetNumSamplesLoaded.num_samples < 100000 then 20000 # Charlie
+                                                                 else 34000
```
so we would be in the 34000 case here... do we think that will actually work Cromwell-wise?

KevinCLydon replied:
That's a good question.  It's below the Terra maximum so it works there, but I haven't tried directly with a Cromwell instance.  I assume it's dependent on how the Cromwell instance is configured.

This logic would really benefit from a rework to be more robust.  From the tests I did, it seems like the number of shards needed is actually dependent on a combination of the number of samples you're extracting, the number of the samples in the full GVS callset, and the size of the interval list, and each of those has a pretty significant effect.  For example, I never successfully ran a whole genome extract for 100k or more samples (even with 34k shards), but I got up to 250k samples with the acaf_threshold interval list, and that was only using 20k shards (and it probably would've worked with even less).  I'm just not sure how best to apply that knowledge to improving this.

KevinCLydon replied:
I'd be happy to share any data I can to help improve it

mcovarr commented on scripts/variantstore/wdl/GvsUtils.wdl:
```diff
@@ -233,6 +233,101 @@ task SplitIntervals {
+  }
+
+  Int disk_size = select_first([split_intervals_disk_size_override, 10])
+  Int disk_memory = select_first([split_intervals_mem_override, 16])
```
I think `disk_memory` is maybe supposed to be `memory_size`?

KevinCLydon replied:
Actually yes, it probably should be.  I just copied that from SplitIntervals.  Should I change it in both places?

mcovarr replied:
yes please! 🙏

mcovarr commented on scripts/variantstore/wdl/MergePgenHierarchical.wdl:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    runtime {
+        docker: "ubuntu:22.04"
```
Is this code very particular about being pinned to this version of Ubuntu or could we specify `basic_docker` as defined by `GvsUtils.wdl`?

KevinCLydon replied:
I can make that change.  One or two of the tasks in that file need that specific docker because of some bash syntax I'm using, but I'm pretty sure that task does not.

mcovarr commented on src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/GuideToPGENExtract.md:
```diff
@@ -0,0 +1,85 @@
+### Part 1: PGEN-JNI
+The PGEN-JNI library was written by Chris Norman of the GATK Engine Team and lives [here](https://github.com/broadinstitute/pgen-jni).  It is written primarily in C++ for performance purposes and also compatibility with the pgenlib library (part of the [plink repo](https://github.com/chrchang/plink-ng/tree/master)).  It builds on top of pgenlib to provide a writer for creating PGEN files and writ ...
+
+PGEN-JNI is compatible with Linux and macOS.
```
Though ["MacOS aarm64 (sic) is not yet supported"](https://github.com/broadinstitute/pgen-jni?tab=readme-ov-file#supported-runtime-platforms), which I believe is what most of us are using.

KevinCLydon replied:
Ooo, good point.  I'll update that

mcovarr reviewed (commented):
First pass -- do you know if the tests are actually executing? There seem to be a ton of warnings about lines in the tests not being covered by tests... 🤔

KevinCLydon commented:
> First pass -- do you know if the tests are actually executing? There seem to be a ton of warnings about lines in the tests not being covered by tests... 🤔

That's a good question, actually.  I assume they're running, but I can look into it.  The codecov results are weird.  A couple of them are showing some exception handling code that's not covered (which I can write a couple more tests for if we want), but it's also calling out the test file, which doesn't make a lot of sense to me.

mcovarr reviewed (commented):
Another round, now I'm going to try to actually use this! 😄

mcovarr commented on src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/GuideToPGENExtract.md:
```diff
@@ -0,0 +1,85 @@
+### Part 1: PGEN-JNI
+The PGEN-JNI library was written by Chris Norman of the GATK Engine Team and lives [here](https://github.com/broadinstitute/pgen-jni).  It is written primarily in C++ for performance purposes and also compatibility with the pgenlib library (part of the [plink repo](https://github.com/chrchang/plink-ng/tree/master)).  It builds on top of pgenlib to provide a writer for creating PGEN files and writ ...
+
+PGEN-JNI is compatible with Linux and macOS (only Intel x64 processors, not ARM).
```
```suggestion
PGEN-JNI is compatible with Linux and macOS (only Intel x86 processors, not ARM).
```

KevinCLydon replied:
Oops, thank you for catching that

KevinCLydon commented:
> The four test failures in `ExtractCohortToPgenTest` appear to be real:
> 
> ```
> 2024-03-12T20:53:02.3169070Z Gradle suite > Gradle test > org.broadinstitute.hellbender.tools.gvs.extract.ExtractCohortToPgenTest > testFinalVQSRLitePgenfromRangesAvro �[31mFAILED�[39m�[0K
> 2024-03-12T20:53:02.3171276Z     java.lang.AssertionError: expected [-1] but found [13570]�[0K
> 2024-03-12T20:53:02.3172934Z         at org.testng.Assert.fail(Assert.java:97)
> 2024-03-12T20:53:02.3173927Z         at org.testng.Assert.assertEqualsImpl(Assert.java:136)
> 2024-03-12T20:53:02.3174922Z         at org.testng.Assert.assertEquals(Assert.java:118)
> 2024-03-12T20:53:02.3175853Z         at org.testng.Assert.assertEquals(Assert.java:729)
> 2024-03-12T20:53:02.3176775Z         at org.testng.Assert.assertEquals(Assert.java:739)
> 2024-03-12T20:53:02.3178703Z         at org.broadinstitute.hellbender.tools.gvs.extract.ExtractCohortToPgenTest.testFinalVQSRLitePgenfromRangesAvro(ExtractCohortToPgenTest.java:78)
> ```

That's weird, because it looks like it's succeeding in the other test tasks where it's running.  The place it's failing is an extremely simple equality assertion of two compressed files, though.  I wonder if there's something about operating system differences that can change the compression slightly.  I'll see if I can find a better way to do that check

KevinCLydon commented:
@mcovarr Follow-up on the codecov thing.  Apparently codecov is just not configured correctly?  I asked at an Engine Team meeting and it looks like we're just not excluding test files for some reason, so that explains why it's calling out my test file as not covered.  We could change it to ignore that directory, since it makes sense to not check for test coverage on test files, but I'm hesitant to include it here because it's sort of outside the scope of this PR.

mcovarr commented:
Thanks for inquiring on the codecov issue! I totally agree that fixing that is outside the scope of this PR if the root cause is unrelated to your changes (though it would be lovely if someone could fix that in a separate PR on the main branch 🙂).

mcovarr reviewed (commented):
Hi Kevin, our team would like to get this merged into `ah_var_store` soon per VS-1254. I'm aware of only a handful of outstanding issues:

- The failing PGEN tests. I'm happy to help here in any way I can though right now I don't have a sense of what could be causing this beyond the platform differences you suggested.
- The `10` vs `10.0` change we discussed recently to avoid division by zero.
- We'll want to merge / rebase from `ah_var_store` and then build a new GATK Docker image which would be entered into `GetToolVersions` in `GvsUtils.wdl`. I'm happy to take on building this image once the merge / rebase is ready.

KevinCLydon commented:
@mcovarr I committed your suggestions, fixed the failing tests, and rebased.  Should be ready for you to build the docker image and do whatever final checks are necessary.

mcovarr reviewed (approved):
Thanks Kevin! I've built the updated GATK Docker so I would change line 76 of `GvsUtils.wdl` to say

```
    String gatk_docker = "us.gcr.io/broad-dsde-methods/broad-gatk-snapshots:varstore_2024_03_19_dfd45f6"
```

KevinCLydon commented:
> Thanks Kevin! I've built the updated GATK Docker so I would change line 76 of `GvsUtils.wdl` to say
> 
> ```
>     String gatk_docker = "us.gcr.io/broad-dsde-methods/broad-gatk-snapshots:varstore_2024_03_19_dfd45f6"
> ```

Okay awesome, I just pushed that change.  Am I clear to merge, or is there anything else that needs doing before that?

RoriCremer commented on .dockstore.yml:
```diff
@@ -413,3 +413,21 @@ workflows:
+     primaryDescriptorPath: /scripts/variantstore/wdl/GvsExtractCallsetPgen.wdl
+     filters:
+       branches:
+         - kl_pgen_extract_for_variants_pr
```
are we going to need ah_var_store branches?

mcovarr replied:
First we'll need `EchoCallset` for `GvsExtractCallsetPgenMerged`, but we can add that when we merge this from `ah_var_store` to `EchoCallset` (separate from this PR).

KevinCLydon replied:
Ooo good catch!  Thank you for pointing that out.  I completely forgot to update that

mcovarr commented:
Hi Kevin, I asked my team to look this over once more before we merge. I'd say you're good to go with one more 👍

RoriCremer commented on src/main/java/org/broadinstitute/hellbender/tools/gvs/extract/ExtractCohortToPgen.java:
```diff
@@ -0,0 +1,272 @@
+            doc = "Preserves phasing in the output PGEN file if phasing is present in the source genotypes.",
+            optional = true
+    )
+    private boolean preservePhasing = false;
```
wouldn't we want phasing to default to being on?

KevinCLydon replied:
This is a good question.  It's disabled by default in the PGEN writer library.  I think it maybe makes the files bigger because it changes how the data is stored.  I could change it here, maybe default it to preserve phasing, and have a flag called "--ignore-phasing" or something.  Or I could change it in the WDL to default to preserve phasing.  What do you think?

RoriCremer replied:
ah the size is something I hadn't thought about (silly me) and that's enough to convince me

mcovarr reviewed (approved):
Hi Kevin, everything looks good from the Variants team perspective so I think you're good to merge. Thank you again for all your hard work on this! 👍

KevinCLydon commented:
Thank you!
